compiler/rustc_interface/src/interface.rs

```diff
@@ -117,19 +121,20 @@ pub fn parse_cfgspecs(
                     error!(r#"expected `key` or `key="value"`"#);
                 }
             })
-            .collect::<CrateConfig>();
-        cfg.into_iter().map(|(a, b)| (a.to_string(), b.map(|b| b.to_string()))).collect()
+            .collect::<Cfg<String>>()
     })
 }
 
 /// Converts strings provided as `--check-cfg [specs]` into a `CheckCfg`.
-pub fn parse_check_cfg(handler: &EarlyErrorHandler, specs: Vec<String>) -> CheckCfg {
+pub fn parse_check_cfg(handler: &EarlyErrorHandler, specs: Vec<String>) -> CheckCfg<String> {
+    // The comment about `SessionGlobals` and symbols in `parse_cfg` above
+    // applies here too.
     rustc_span::create_default_session_if_not_set_then(move |_| {
         // If any --check-cfg is passed then exhaustive_values and exhaustive_names
         // are enabled by default.
-        let exhaustive_names = !specs.is_empty();
-        let exhaustive_values = !specs.is_empty();
-        let mut check_cfg = CheckCfg { exhaustive_names, exhaustive_values, ..CheckCfg::default() };
+        let mut exhaustive_names = !specs.is_empty();
+        let mut exhaustive_values = !specs.is_empty();
+        let mut expecteds = FxHashMap::default();
```

Comment:
This makes me a bit sad. I found operating on a `CheckCfg` struct to be clearer.
But I suppose it's for the great good.

Reply:
I think I would prefer keeping around the struct from the start, too. It may be good to add some clearer methods instead (e.g. deduplicating the `old_syntax == None` and immediate overwrite check, or adding methods for inserting new elements to the `expecteds` field).

Reply:
I'll just remove the last commit, it's not that important.